FerramAerospaceResearch/FARAeroComponents/FARAeroSection.cs

```diff
@@ -285,182 +286,103 @@ public void ClearAeroSection()
-
-            Vector3 localNormalForceVec = Vector3.ProjectOnPlane(-velLocalNorm, xRefVector).normalized;
 
-            double cosAoA = Vector3.Dot(xRefVector, velLocalNorm);
-            double cosSqrAoA = cosAoA * cosAoA;
-            double sinSqrAoA = Math.Max(1 - cosSqrAoA, 0);
-            double sinAoA = Math.Sqrt(sinSqrAoA);
-            double sin2AoA = 2 * sinAoA * Math.Abs(cosAoA);
-            double cosHalfAoA = Math.Sqrt(0.5 + 0.5 * Math.Abs(cosAoA));
-
-
-            double nForce = 0;
-            nForce = potentialFlowNormalForce * Math.Sign(cosAoA) * cosHalfAoA * sin2AoA;  //potential flow normal force
-            if (nForce < 0)     //potential flow is not significant over the rear face of things
-                nForce = 0;
-            //if (machNumber > 3)
-            //    nForce *= 2d - machNumber * 0.3333333333333333d;
-
-            float normalForceFactor = Math.Abs(Vector3.Dot(localNormalForceVec, nRefVector));
-            normalForceFactor *= normalForceFactor;
-
-            normalForceFactor = invFlatnessRatio * (1 - normalForceFactor) + flatnessRatio * normalForceFactor;     //accounts for changes in relative flatness of shape
-
-
-            float crossFlowMach, crossFlowReynolds;
-            crossFlowMach = machNumber * (float)sinAoA;
-            crossFlowReynolds = reynoldsPerUnitLength * diameter * (float)sinAoA / normalForceFactor;
-
-            nForce += viscCrossflowDrag * sinSqrAoA * CalculateCrossFlowDrag(crossFlowMach, crossFlowReynolds);            //viscous crossflow normal force
-
-            nForce *= normalForceFactor;
-
-            double xForce = -skinFrictionForce * Math.Sign(cosAoA) * cosSqrAoA;
-            double localVelForce = xForce * pseudoKnudsenNumber;
-            xForce -= localVelForce;
-
-            localVelForce = Math.Abs(localVelForce);
-
-            float moment = (float)(cosAoA * sinAoA);
-            float dampingMoment = 4f * moment;
-
-
-            if (cosAoA > 0)
+            public Vector3 LocalVelocity(PartData pd)
             {
-                xForce += cosSqrAoA * xForceAoA0;
-                float momentFactor;
-                if (machNumber > 6)
-                    momentFactor = hypersonicMomentForward;
-                else if (machNumber < 0.6)
-                    momentFactor = 0.6f * hypersonicMomentBackward;
-                else
-                {
-                    float tmp = (-0.185185185f * machNumber + 1.11111111111f);
-                    momentFactor = tmp * hypersonicMomentBackward * 0.6f + (1 - tmp) * hypersonicMomentForward;
-                }
-                //if (machNumber < 1.5)
-                //    momentFactor += hypersonicMomentBackward * (0.5f - machNumber * 0.33333333333333333333333333333333f) * 0.2f;
-
-                moment *= momentFactor;
-                dampingMoment *= momentFactor;
+                return pd.aeroModule.part.partTransform.InverseTransformVector(worldVel);
```

Comment:
This throws NREs when the vehicle breaks up (if MechJeb is installed). Maybe add another method (something like `IsValid(PartData pd)`?) to the interface which checks if the current `PartData pd` is valid. So for flight forces it would be `(object)pd.aeroModule != null` and for simulation in addition check if `pd.aeroModule.part != null` and `pd.aeroModule.part.partTransform != null`.

Also, consider caching the return value since `Transform.InverseTransformVector` calls an internal method (managed to native) which is slow (compared to usual C# calls).

Reply:
There's already a check `(object)pd.aeroModule != null` in the calling context; see line 389 of FARAeroSection.cs. I think it's okay if simulation produces a 0 for that part in the latter two cases (and I'm not overly concerned with simulation performance; I can basically do a full wind tunnel like test in <0.5s), so to minimize vcalls in the flight code I'll put that into SimulationContext only.